desc/perturbations.py

```diff
@@ -588,39 +590,56 @@ def optimal_perturb(  # noqa: C901 - FIXME: break this up into simpler pieces
         f = objective_f.compute(xf)
         g = objective_g.compute(xg)
 
-        # 1st partial derivatives of f objective wrt both x and c
+        # 1st partial derivatives of f objective wrt x
         if verbose > 0:
             print("Computing df")
         timer.start("df computation")
         Fx = objective_f.jac(xf)
         Fx = align_jacobian(Fx, objective_f, objective_g)
-        Fx_reduced = Fx[:, unfixed_idx] @ Z
-        Fc = Fx @ dxdc
-        if cutoff is None:
-            cutoff = np.finfo(Fx_reduced.dtype).eps * np.max(Fx_reduced.shape)
-        Fx_reduced_inv = np.linalg.pinv(Fx_reduced, rcond=cutoff)
         timer.stop("df computation")
         if verbose > 1:
             timer.disp("df computation")
 
-        # 1st partial derivatives of g objective wrt both x and c
+        # 1st partial derivatives of g objective wrt x
         if verbose > 0:
             print("Computing dg")
         timer.start("dg computation")
         Gx = objective_g.jac(xg)
         Gx = align_jacobian(Gx, objective_g, objective_f)
-        Gx_reduced = Gx[:, unfixed_idx] @ Z
-        Gc = Gx @ dxdc
         timer.stop("dg computation")
         if verbose > 1:
             timer.disp("dg computation")
 
-        GxFx = Gx_reduced @ Fx_reduced_inv
+        # projections onto optimization space
+        Fx_reduced = Fx[:, unfixed_idx] @ Z
+        Gx_reduced = Gx[:, unfixed_idx] @ Z
+        Fc = Fx @ dxdc
+        Gc = Gx @ dxdc
+
+        # some scaling to improve conditioning and rescale trust region
+        wf, _ = compute_jac_scale(Fx_reduced)
+        wg, _ = compute_jac_scale(Gx_reduced)
+        wx = wf + wg
+        Fxh = Fx_reduced * wx
+        Gxh = Gx_reduced * wx
+        if cutoff is None:
+            cutoff = np.finfo(Fx_reduced.dtype).eps * np.max(Fx_reduced.shape)
+        uf, sf, vtf = np.linalg.svd(Fxh, full_matrices=False)
+        sf += sf[-1]  # add a tiny bit of regularization
```

Comment:
Adding a constant to the singular values reduces the condition number. Ideally we'd use a trust region thing but that's tricky to do for both x and c, and emperically I found that the trust region step was usually adding something on the order of the smallest singular value of the rescaled jacobian, so that's what I used here.